README


This archive contains the distribution Dancer-Plugin-Locale-Wolowitz,
version 0.140120:

  Intenationalization for Dancer

This software is copyright (c) 2014 by Natal Ngétal.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.009.



Changes
Revision history for Dancer-Plugin-Locale-Wolowitz

0.140120  2014-01-12 01:53:54CET+0100 Europe/Paris

    Few cosmetic things.
    Fix unitialized value in substitution when accept_language is empty, thanks
    to isync for the patch.

0.122470  2012-09-03 23:32:10 Europe/Paris

    Add few tests.
    Add exception if translate directory not found.

0.121720  2012-06-20 23:31:40 Europe/Paris

    Persistent Wolowitz obj so json isn't parsed over and over again.
    Don't need use a session engine.

0.121391  2012-05-18 23:39:44 Europe/Paris

    Fix global variable.
    Fix tests.

0.121390  2012-05-18 21:24:33 Europe/Paris

    Initial version developed.


t/lib/TestApp.pm
#
# This file is part of Dancer-Plugin-Locale-Wolowitz
#
# This software is copyright (c) 2014 by Natal Ngétal.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package t::lib::TestApp;

use Dancer;
use Dancer::Plugin::Locale::Wolowitz;

get '/' => sub {
    my $tr = loc('welcome');
    return $tr;
};

get '/no_key' => sub {
    my $tr = loc('hello');
    return $tr;
};

get '/complex_key' => sub {
    my $path = setting('appdir');
    my $tr   = loc('path_not_found %1', [$path]);

    return $tr;
};

1;


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.009.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Intenationalization for Dancer",
  "dist_author" => [
    "Natal Ng\x{e9}tal"
  ],
  "dist_name" => "Dancer-Plugin-Locale-Wolowitz",
  "dist_version" => "0.140120",
  "license" => "perl",
  "module_name" => "Dancer::Plugin::Locale::Wolowitz",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dancer" => 0,
    "Dancer::Exception" => 0,
    "Dancer::Plugin" => 0,
    "Locale::Wolowitz" => 0,
    "perl" => "5.010",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => [],
  "test_requires" => {
    "Dancer::Test" => 0,
    "Test::More" => 0,
    "lib" => 0
  }
);


my %fallback_build_requires = (
  "Dancer::Test" => 0,
  "Module::Build" => "0.3601",
  "Test::More" => 0,
  "lib" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer/Plugin/Locale/Wolowitz.pm
#
# This file is part of Dancer-Plugin-Locale-Wolowitz
#
# This software is copyright (c) 2014 by Natal Ngétal.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Dancer::Plugin::Locale::Wolowitz;
{
  $Dancer::Plugin::Locale::Wolowitz::VERSION = '0.140120';
}

use strict;
use warnings;

use 5.010;

use Dancer ':syntax';
use Dancer::Plugin;
use Dancer::Exception qw(:all);

use Locale::Wolowitz;

#ABSTRACT: Intenationalization for Dancer

my $w;

#Register exception
register_exception('DirectoryNotFound',
    message_pattern => "Not found directory: %s"
);


add_hook(
    before_template => sub {
        my $tokens = shift;

        $tokens->{l} = sub { _loc(@_); };
    }
);

register loc => sub {
    _loc(@_);
};

sub _loc {
    my ( $str, $args ) = @_;

    $w       = Locale::Wolowitz->new(_path_directory_locale()) unless defined($w);
    my $lang = _lang();

    !$args and return $w->loc($str, $lang);

    my $msg = $w->loc($str, $lang, map($w->loc($_, $lang), @{$args}));

    return $msg;
}

sub _path_directory_locale {
    my $settings = plugin_setting;
    my $path     = $settings->{locale_path_directory}
        // Dancer::FileUtils::path(setting('appdir'), 'i18n');

    if ( ! -d $path ) {
        raise DirectoryNotFound => $path;
    }

    return $path;
}

sub _lang {
    my $settings     = plugin_setting;
    my $lang_session = $settings->{lang_session} || 'lang';
    my $lang;

    # don't force the user to store lang in a session
    if ( setting('session') ) {
        my $session_language = session $lang_session;

        if ( !$session_language ) {
            $lang = _detect_lang_from_browser();

            session $lang_session => $lang;
            return $lang;
        }
        else {
            return $session_language;
        }
    }

    $lang = _detect_lang_from_browser();

    return $lang;
}

sub _detect_lang_from_browser {
    my $lang = request->accept_language;

    return unless $lang;

    $lang =~ s/-\w+//g;
    $lang = (split(/,\s*/,$lang))[0] if $lang =~ /,/;

    return $lang;
}

register_plugin;

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Dancer::Plugin::Locale::Wolowitz - Intenationalization for Dancer

=head1 VERSION

version 0.140120

=head1 SYNOPSIS

    use Dancer ':syntax';
    use Dancer::Plugin::Locale::Wolowitz;

    get '/' => sub {
        template index;
    }

=head1 DESCRIPTION

Provides an easy way to translate your application. This module relies on L<Locale::Wolowitz>, please consult the documentation of Locale::Wolowitz.

=head1 METHODS

=head2 loc

    loc('Welcome');
    loc('View %1', ['Country'])
or
    <% l('Welcome') %>
    <% l('View %1', ['Country']) %>

Translated to the requested language, if such a translation exists, otherwise
no traslation occurs.

    input: (Str): Key translate
           (Arrayref): Arguments are injected to the placeholders in the string
    output: (Str): Translated to the requested language

=head1 CONFIGURATION

  plugins:
    Locale::Wolowitz:
      lang_session: "lang"
      locale_path_directory: "i18n"

=head1 CONTRIBUTING

This module is developed on Github at:

L<http://github.com/hobbestigrou/Dancer-Plugin-Locale-Wolowitz>

=head1 ACKNOWLEDGEMENTS

Thanks to Ido Perlmuter to Locale::Wolowitz

=head1 BUGS

Please report any bugs or feature requests in github.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Dancer::Plugin::Locale::Wolowitz

=head1 SEE ALSO

L<Dancer>
L<Locale::Wolowitz>

=head1 AUTHOR

Natal Ngétal

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Natal Ngétal.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.009.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Intenationalization for Dancer",
  "AUTHOR" => "Natal Ng\x{e9}tal",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Dancer-Plugin-Locale-Wolowitz",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Dancer::Plugin::Locale::Wolowitz",
  "PREREQ_PM" => {
    "Dancer" => 0,
    "Dancer::Exception" => 0,
    "Dancer::Plugin" => 0,
    "Locale::Wolowitz" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Dancer::Test" => 0,
    "Test::More" => 0,
    "lib" => 0
  },
  "VERSION" => "0.140120",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Dancer" => 0,
  "Dancer::Exception" => 0,
  "Dancer::Plugin" => 0,
  "Dancer::Test" => 0,
  "Locale::Wolowitz" => 0,
  "Module::Build" => "0.3601",
  "Test::More" => 0,
  "lib" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



